File: agents/windows/src/N3NAgent.Core/Capabilities/ShellCapability.cs
using System.Diagnostics;

namespace N3NAgent.Core.Capabilities;

/// <summary>
/// Capability for executing shell commands.
/// </summary>
public class ShellCapability : ICapability
{
    public string Id => "shell.execute";
    public string Name => "Shell Execute";
    public string Description => "Execute shell commands and return output";

    private const int MaxTimeout = 300; // seconds

    public async Task<object?> ExecuteAsync(Dictionary<string, object> args)
    {
        if (!args.TryGetValue("command", out var commandObj) || commandObj == null)
            throw CapabilityException.InvalidArguments("Missing 'command' argument");

        var command = commandObj.ToString()!;
        var timeout = args.TryGetValue("timeout", out var t) ? Math.Min(Convert.ToInt32(t), MaxTimeout) : 60;
        var workingDirectory = args.TryGetValue("workingDirectory", out var wd) ? wd.ToString() : null;

        return await ExecuteCommand(command, timeout, workingDirectory);
    }

    private async Task<Dictionary<string, object>> ExecuteCommand(string command, int timeout, string? workingDirectory)
    {
        using var process = new Process();

        process.StartInfo = new ProcessStartInfo
        {
            FileName = "cmd.exe",
            Arguments = $"/c {command}",
            UseShellExecute = false,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            CreateNoWindow = true,
            WorkingDirectory = workingDirectory ?? Environment.CurrentDirectory
        };

        var stdout = new StringBuilder();
        var stderr = new StringBuilder();

        process.OutputDataReceived += (_, e) =>
        {
            if (e.Data != null) stdout.AppendLine(e.Data);
        };

        process.ErrorDataReceived += (_, e) =>
        {
            if (e.Data != null) stderr.AppendLine(e.Data);
        };

        try
        {
            process.Start();
            process.BeginOutputReadLine();
            process.BeginErrorReadLine();

            var completed = await Task.Run(() => process.WaitForExit(timeout * 1000));

            if (!completed)
            {
                process.Kill();
                throw CapabilityException.ExecutionFailed("Command timed out");
            }

            return new Dictionary<string, object>
            {
                ["exitCode"] = process.ExitCode,
                ["stdout"] = stdout.ToString(),
                ["stderr"] = stderr.ToString(),
                ["success"] = process.ExitCode == 0
            };
        }
        catch (Exception ex) when (ex is not CapabilityException)
        {
            throw CapabilityException.ExecutionFailed($"Failed to execute command: {ex.Message}");
        }
    }
}

File: agents/macos/Sources/N3NAgentCore/Capabilities/ShellCapability.swift
import Foundation

/// Capability for executing shell commands.
public final class ShellCapability: Capability {

    public let id = "shell.execute"
    public let name = "Shell Execute"
    public let description = "Execute shell commands and return output"

    /// Maximum execution time in seconds
    private let maxTimeout: TimeInterval = 300

    public init() {}

    public func execute(args: [String: Any]) async throws -> Any? {
        guard let command = args["command"] as? String else {
            throw CapabilityError.invalidArguments("Missing 'command' argument")
        }

        let timeout = min(args["timeout"] as? TimeInterval ?? 60, maxTimeout)
        let workingDirectory = args["workingDirectory"] as? String
        let environment = args["environment"] as? [String: String]

        return try await executeCommand(
            command: command,
            timeout: timeout,
            workingDirectory: workingDirectory,
            environment: environment
        )
    }

    private func executeCommand(
        command: String,
        timeout: TimeInterval,
        workingDirectory: String?,
        environment: [String: String]?
    ) async throws -> [String: Any] {
        let process = Process()
        let stdoutPipe = Pipe()
        let stderrPipe = Pipe()

        process.executableURL = URL(fileURLWithPath: "/bin/zsh")
        process.arguments = ["-c", command]
        process.standardOutput = stdoutPipe
        process.standardError = stderrPipe

        if let workingDirectory = workingDirectory {
            process.currentDirectoryURL = URL(fileURLWithPath: workingDirectory)
        }

        if let environment = environment {
            var env = ProcessInfo.processInfo.environment
            for (key, value) in environment {
                env[key] = value
            }
            process.environment = env
        }

        // Start process
        do {
            try process.run()
        } catch {
            throw CapabilityError.executionFailed("Failed to start process: \(error.localizedDescription)")
        }

        // Create timeout task
        let timeoutTask = Task {
            try await Task.sleep(nanoseconds: UInt64(timeout * 1_000_000_000))
            if process.isRunning {
                process.terminate()
            }
        }

        // Wait for process
        process.waitUntilExit()
        timeoutTask.cancel()

        // Read output
        let stdoutData = stdoutPipe.fileHandleForReading.readDataToEndOfFile()
        let stderrData = stderrPipe.fileHandleForReading.readDataToEndOfFile()

        let stdout = String(data: stdoutData, encoding: .utf8) ?? ""
        let stderr = String(data: stderrData, encoding: .utf8) ?? ""

        return [
            "exitCode": process.terminationStatus,
            "stdout": stdout,
            "stderr": stderr,
            "success": process.terminationStatus == 0
        ]
    }
}
